src/swarm-topography/Swarm/Game/Scenario/Topography/Grid.hs

```diff
@@ -2,63 +2,84 @@
 -- SPDX-License-Identifier: BSD-3-Clause
 module Swarm.Game.Scenario.Topography.Grid (
   Grid (..),
+  NonEmptyGrid (..),
   gridToVec,
-  mapIndexedMembers,
+  zipNumberedNE,
+  mapWithCoordsNE,
+  mapWithCoords,
   allMembers,
-  mapRows,
+  mapRowsNE,
   getRows,
   mkGrid,
 )
 where
 
 import Data.Aeson (ToJSON (..))
+import Data.Foldable qualified as F
 import Data.List.NonEmpty (NonEmpty)
 import Data.List.NonEmpty qualified as NE
 import Data.Maybe (fromMaybe)
 import Data.Semigroup
 import Data.Vector qualified as V
+import GHC.Generics (Generic)
 import Swarm.Game.World.Coords
 import Prelude hiding (zipWith)
 
+newtype NonEmptyGrid c = NonEmptyGrid (NonEmpty (NonEmpty c))
+  deriving (Generic, Show, Eq, Functor, Foldable, Traversable, ToJSON)
```

Comment:
Seeing the changes mixed with this refactoring takes a lot of work. 🙁

Reply:
Good point, I should have made a separate PR for that.